xwiki-platform-core/xwiki-platform-notifications/xwiki-platform-notifications-filters/xwiki-platform-notifications-filters-default/src/main/java/org/xwiki/notifications/filters/migration/R140400000XWIKI15460DataMigration.java

```diff
@@ -0,0 +1,121 @@
+import org.xwiki.stability.Unstable;
+import org.xwiki.wiki.descriptor.WikiDescriptorManager;
+import org.xwiki.wiki.manager.WikiManagerException;
+
+import com.xpn.xwiki.XWikiContext;
+import com.xpn.xwiki.store.migration.DataMigrationException;
+import com.xpn.xwiki.store.migration.XWikiDBVersion;
+import com.xpn.xwiki.store.migration.hibernate.AbstractHibernateDataMigration;
+
+/**
+ * Cleanup the notification filters preferences remaining on the main wiki from previously removed sub-wikis. Note: this
+ * class is named {@code R131006000XWIKI1546} in branch {@code 13.10.6+}.
+ *
+ * @version $Id$
+ * @see <a href="https://jira.xwiki.org/browse/XWIKI-15460">XWIKI-15460: Notification filter preferences are not cleaned
+ *     when a wiki is deleted</a>
+ * @since 14.4
+ * @since 13.10.6
+ */
+@Component
+@Singleton
+@Named("R140400000XWIKI1546")
+@Unstable
+public class R140400000XWIKI15460DataMigration extends AbstractHibernateDataMigration
+{
+    @Inject
+    private WikiDescriptorManager wikiDescriptorManager;
+
+    @Inject
+    private NotificationFilterPreferenceStore store;
+
+    @Inject
+    private Logger logger;
+
+    @Override
+    public XWikiDBVersion getVersion()
+    {
+        return new XWikiDBVersion(140400000);
+    }
+
+    @Override
+    public String getDescription()
+    {
+        return "Remove the notification filters preferences remaining from removed sub-wikis.";
+    }
+
+    @Override
+    protected void hibernateMigrate() throws DataMigrationException
+    {
+        XWikiContext context = getXWikiContext();
+        if (!Objects.equals(context.getWikiId(), context.getMainXWiki())) {
+            this.logger.info("Skipping, this migration only applies to the main wiki.");
+            return;
+        }
+
+        int version = this.manager.get().getDBVersion().getVersion();
+        if (version >= 131006000 && version < 140000000) {
+            this.logger.info("Skipping, this migration has already been performed in 13.10.6+.");
+            return;
+        }
+
+        try {
+            Collection<String> wikiIds = this.wikiDescriptorManager.getAllIds();
+            int limit = 1000;
+            int offset = 0;
+            Set<NotificationFilterPreference> allNotificationFilterPreferences = this.store
+                .getPaginatedFilterPreferences(limit, offset);
+            while (!allNotificationFilterPreferences.isEmpty()) {
+                // We count the deleted filter preferences and adapt the next offset accordingly.
+                int removed = 0;
+                for (NotificationFilterPreference filterPreference : allNotificationFilterPreferences) {
+                    boolean isFromExistingWiki = wikiIds.stream().anyMatch(filterPreference::isFromWiki);
+                    if (!isFromExistingWiki) {
+                        removed++;
+                        this.store.deleteFilterPreference(filterPreference);
```

Comment:
Not a fan of deleting each preference one by one, I feel it would be better to gather all the not existing wikis references in the preference you are to load and then just do a single delete (or maybe one per wiki to cleanup if it's too complex).

Reply:
I considered this option but I got afraid of the memory usage if the number of filter preferences to retain before delete is large.

Reply:
You don't need to keep any preferences, just the referenced wikis.

Reply:
Ah right, that should work indeed.

Reply:
We currently don't have a way to extract the wikiId from a filter preference, but that shouldn't be difficult to add.

Reply:
Yes, it's not so different from isFromWiki (which you might actually not need anymore unless you are using it somewhere else).

Reply:
It's still used for cache invalidation.